sdk/core/System.ClientModel/src/Message/PipelineMessage.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Threading;
+
+namespace System.ClientModel.Primitives;
+
+// TODO: MessageClassifier and RequestOptions will come in a later PR.
+
+public class PipelineMessage : IDisposable
+{
+    private ArrayBackedPropertyBag<ulong, object> _propertyBag;
+    private bool _disposed;
+
+    protected internal PipelineMessage(PipelineRequest request)
+    {
+        Argument.AssertNotNull(request, nameof(request));
+
+        Request = request;
+        _propertyBag = new ArrayBackedPropertyBag<ulong, object>();
+    }
+
+    public PipelineRequest Request { get; }
+
+    public PipelineResponse? Response
+    {
+        get;
+
+        // Set internally by the transport.
+        protected internal set;
+    }
+
+    #region Pipeline invocation options
+
+    public CancellationToken CancellationToken { get; set; }
+
+    public bool TryGetProperty(Type type, out object? value) =>
+        _propertyBag.TryGetValue((ulong)type.TypeHandle.Value, out value);
+
+    public void SetProperty(Type type, object value) =>
+        _propertyBag.Set((ulong)type.TypeHandle.Value, value);
+
+    #endregion
+
+    #region Meta-data for pipeline processing
+
+    internal int RetryCount { get; set; }
+
+    #endregion
+
+    #region Per-request pipeline
+
+    internal bool CustomRequestPipeline =>
+        PerCallPolicies is not null || PerTryPolicies is not null;
+
+    internal PipelinePolicy[]? PerCallPolicies { get; set; }
+
+    internal PipelinePolicy[]? PerTryPolicies { get; set; }
+
+    internal PipelinePolicy[]? BeforeTransportPolicies { get; set; }
+
+    #endregion
+
+    internal void AssertResponse()
+    {
+        if (Response is null)
+        {
+            throw new InvalidOperationException("Response is not set on message.");
+        }
+    }
+
+    #region IDisposable
+
+    protected virtual void Dispose(bool disposing)
+    {
+        if (disposing && !_disposed)
+        {
+            var request = Request;
```

Comment:
Do we need the intermediate variable?  It doesn't look like `Request` is being set to `null` here nor that we first check `request` to ensure it is not `null`.

Reply:
IIRC this is boilerplate implementation and done this way because of thread safety.  Are you concerned about perf?  I think because the intermediate variable goes on the stack it's not too big a perf issue.

Reply:
More readability and understandability of code.   The current form makes one question "why" and can lead to incorrect conclusions about how Request is managed.